src/libraries/System.Private.Uri/src/System/UriHelper.cs

```diff
@@ -118,26 +118,79 @@ public static unsafe string SpanToLowerInvariantString(ReadOnlySpan<char> span)
             return true;
         }
 
-        internal static string EscapeString(string stringToEscape, bool checkExistingEscaped, SearchValues<char> noEscape)
+        public static bool TryEscapeDataString(ReadOnlySpan<char> charsToEscape, Span<char> destination, out int charsWritten)
+        {
+            if (destination.Length < charsToEscape.Length)
+            {
+                charsWritten = 0;
+                return false;
+            }
+
+            int indexOfFirstToEscape = charsToEscape.IndexOfAnyExcept(Unreserved);
+            if (indexOfFirstToEscape < 0)
+            {
+                // Nothing to escape, just copy the original chars.
+                charsToEscape.CopyTo(destination);
+                charsWritten = charsToEscape.Length;
+                return true;
+            }
+
+            var vsb = new ValueStringBuilder(stackalloc char[Uri.StackallocThreshold]);
+
+            // This may throw for very large inputs. ¯\_(ツ)_/¯
+            vsb.EnsureCapacity(charsToEscape.Length - indexOfFirstToEscape);
```

Comment:
@stephentoub should we care about cases like:
a) The input span is larger than max array length, so we can't rent the temporary buffer. But _technically_ the destination might be large enough to hold the entire result.
b) When escaping and the output grows, it grows beyond array max length, but lower than destination length, so it _technically_ could work (just not via `ValueStringBuilder`).
c) When escaping the output grows beyond int32

and/or should we care about which exception we end up throwing in such cases?

Realistically I hope no one would ever hit this.

Reply:
> should we care about

We shouldn't corrupt anything or the like, but if we throw an OutOfMemoryException or the like when close to the Array.MaxLength limit even if we're not technically over it, that's fine, at least for now. If we get feedback that it's an issue, we can tweak it later, but hopefully folks don't have 2B character Uris.